sydent/db/hashing_metadata.py

```diff
@@ -0,0 +1,110 @@
+        """Stores values in the hashing_metadata table under the named columns
+
+        :param names_and_values: Column names and associated values to store
+                                 in the database
+        :type names_and_values: Dict
+        """
+        cur = self.sydent.db.cursor()
+
+        columns = ', '.join(names_and_values.keys())
+        values = ', '.join('?' * len(names_and_values))
+        sql = 'INSERT INTO hashing_metadata ({}) VALUES ({})'.format(columns, values)
+        values = names_and_values.values()
+        cur.execute(sql, values)
+        self.sydent.db.commit()
+
+    def rehash_threepids(self, hashing_function, pepper):
+        """Rehash all 3PIDs using a given hashing_function and pepper
+
+        :param hashing_function: A function with single input and output strings
+        :type hashing_function func(str) -> str
+
+        :param pepper: A pepper to append to the end of the 3PID (after a space) before hashing
+        :type pepper: str
+        """
+        self._rehash_threepids(hashing_function, pepper, "local_threepid_associations")
+        self._rehash_threepids(hashing_function, pepper, "global_threepid_associations")
+
+    def _rehash_threepids(self, hashing_function, pepper, table):
+        """Rehash 3PIDs of a given table using a given hashing_function and pepper
+
+        :param hashing_function: A function with single input and output strings
+        :type hashing_function func(str) -> str
+
+        :param pepper: A pepper to append to the end of the 3PID (after a space) before hashing
+        :type pepper: str
+
+        :param table: The database table to perform the rehashing on
+        :type table: str
+        """
+        cur = self.sydent.db.cursor()
+
+        # Pull items from the database
+        # Medium/address combos are marked as UNIQUE in the database
+        sql = "SELECT medium, address FROM %s" % table
+        res = cur.execute(sql)
+        rows = res.fetchall()
+
+        # Iterate through each medium, address combo, hash it,
+        # and store in the db
+        batch_size = 500
+        while rows:
+            for medium, address in rows[:batch_size]:
+                # Combine the medium, address and pepper together in the
+                # following form: "address medium pepper"
+                # According to MSC2134: https://github.com/matrix-org/matrix-doc/pull/2134
+                combo = "%s %s %s" % (address, medium, pepper)
+
+                # Hash the resulting string
+                result = hashing_function(combo)
+
+                # Save the result to the DB
+                sql = (
+                    "UPDATE %s SET lookup_hash = ? "
+                    "WHERE medium = ? AND address = ?"
+                    % (table)
+                )
+                cur.execute(sql, (result, medium, address))
+
+            # Remove processed items from the list
+            rows = rows[batch_size:]
+
+            self.sydent.db.commit()
```

Comment:
So we end up with rows in the same table hashed with different peppers?

Reply:
Ideally not

Reply:
Well you're committing to the DB :)

Reply:
But with update commands, no?

Reply:
Oh, this has been changed! yay!